### ISSUE TITLE
document the expected rules for partially resolving template argument lists

### ISSUE THREAD

**Issue by zygoloid** [itanium-cxx-abi/cxx-abi#113]
Testcase:

```
template<int A, int ...B> struct X {};

// Everyone agrees that the template arguments here are I (D) J (C...) E E
template<int D, int ...C> void h(X<D, C...>) {}
void i() { h<1, 2, 3, 4>(X<1, 2, 3, 4>()); }

// But everyone agrees that the template arguments here are I (C...) (D) E, with no J...E in sight
template<int D, int ...C> void f(X<C..., D>) {}
void g() { f<4, 1, 2, 3>(X<1, 2, 3, 4>()); }
```

Broadly-speaking, it seems that there are two different ways we could mangle a dependent template argument list:

1) Mangle the source syntax directly, or
2) Resolve the template argument list as much as possible, and mangle the source syntax for whatever is left.

Implementation consensus, per the above example, is that the right answer is (2).

Specifically, I think the rule that implementations are using is approximately: perform a left-to-right traversal of the template parameters and arguments, and for each non-dependent parameter/argument pair, convert each argument to the kind and type of the corresponding parameter. Leave unchanged any value-dependent arguments and arguments whose parameters are dependent. If we reach an argument that is a pack expansion where the corresponding parameter is not a pack, then that and all subsequent arguments are mangled in source form (and `J...E` is not used from that point onwards). If we reach a parameter pack, all subsequent explicitly-specified arguments are assumed to correspond to it and are converted appropriately (if the parameter/argument pair is not dependent) and a `J...E` mangling is used for that pack.

We should decide if that's what we want and document the rule.

**Comment by daveedvdv** [comment]
> On Dec 10, 2020, at 9:21 PM, Richard Smith <notifications@github.com> wrote:
> 
> 
> Testcase:
> 
> template<int A, int ...B> struct X {};
> 
> // Everyone agrees that the template arguments here are I (D) J (C...) E E
> template<int D, int ...C> void h(X<D, C...>) {}
> void i() { h<1, 2, 3, 4>(X<1, 2, 3, 4>()); }
> 
> // But everyone agrees that the template arguments here are I (C...) (D) E, with no J...E in sight
> template<int D, int ...C> void f(X<C..., D>) {}
> void g() { f<4, 1, 2, 3>(X<1, 2, 3, 4>()); }
> Broadly-speaking, it seems that there are two different ways we could mangle a dependent template argument list:
> 
> Mangle the source syntax directly, or
> Resolve the template argument list as much as possible, and mangle the source syntax for whatever is left.
> Implementation consensus, per the above example, is that the right answer is (2).
> 
> Specifically, I think the rule that implementations are using is approximately: perform a left-to-right traversal of the template parameters and arguments, and for each non-dependent parameter/argument pair, convert each argument to the kind and type of the corresponding parameter. Leave unchanged any value-dependent arguments and arguments whose parameters are dependent. If we reach an argument that is a pack expansion where the corresponding parameter is not a pack, then that and all subsequent arguments are mangled in source form (and J...E is not used from that point onwards). If we reach a parameter pack, all subsequent explicitly-specified arguments are assumed to correspond to it and are converted appropriately (if the parameter/argument pair is not dependent) and a J...E mangling is used for that pack.
> 
> We should decide if that's what we want and document the rule.
> 

I actually prefer your option 1, but given that 2 is already agreed on de facto, I suppose we’ll have to codify that.


	Daveed

**Comment by rjmccall** [comment]
> I actually prefer your option 1, but given that 2 is already agreed on de facto, I suppose we’ll have to codify that.

Yeah, I agree.  Option 1 is clearly more consistent with the ABI's overall approach to dependent mangling, but it would be a potentially serious compatibility break at this point, and (possibly by coincidence) for once there's implementation agreement on what to do.

Richard, does this get mangled differently in ADL is in play?  Your example has no dependently-typed arguments, so we can resolve the template immediately, but if there were dependent ADL then we wouldn't know whether we might find function templates later.

**Comment by zygoloid** [comment]
We have implementation divergence:
```
template<bool B, typename T = void> struct enable_if { using type = T; };
template<typename ...> struct and_ { static constexpr bool value = true; };
template<typename T> inline typename enable_if<and_<T, T, T>::value>::type f(T) {}
void g() { f(0); }
```
Here, GCC and ICC mangle the `and_` with a `J...E`, and Clang mangles it without. The above proposal would mean that Clang changes to follow the GCC / ICC mangling; I think that's reasonable. (The above is reduced from libstdc++

Regarding ADL: I think the most reasonable model here is: if the function name is a dependent name, then the arguments are mangled unresolved. If it's not a dependent name, they're mangled resolved. That appears to be what ICC and Clang do, but not what GCC does. Testcase:
```
struct X {};
template<unsigned> int f(X);

template<typename T> void g1(decltype(f<0>(T()))) {}
template<typename T> void g2(decltype(f<0>({}) + T())) {}
template<typename T> void g3(decltype(f<0>(X{}) + T())) {}

void h() {
    // Everyone says the template argument is Li0E
    g1<X>({});
    // ICC and Clang say the template argument is resolved to Lj0E, GCC mangles it as Li0E
    g2<int>({});
    g3<int>({});
}
```
So, it looks like GCC partially follows option 1 for at least some function calls in instantiation-dependent expressions, even if there are no associated namespaces. And Clang partially follows option 1 too for at least some class template-ids in instantiation-dependent expressions.

I still think that option 2 is closer to what implementations are doing most of the time. And I'm not entirely sure that option 1 as originally presented is even permitted by the current language rules, because given:
```
template<int ...> struct X {};
template<int ...A> void f(X<A..., 1 + 2>);
template<int ...A> void f(X<A..., 3>);
```
... the expressions `1 + 2` and `3` are not instantiation-dependent, so I don't think an implementation is permitted to mangle them differently (the two `f`s are required to be redeclarations of the same template). So we are required to partially-resolve template argument lists, to some extent, and implementations do so.

[However, there's implementation divergence there, too: for a call to `f<1, 2>({})`, we get two different manglings:
* `_Z1fIJLi1ELi2EEEv1XIJXspT_ELi3EEE` from GCC / Clang
* `_Z1fIJLi1ELi2EEEv1XIJXT_ELi3EEE` from ICC
As far as I can see, the GCC / Clang mangling is correct.]

**Comment by zygoloid** [comment]
How about this: we mangle the original source syntax for the template arguments (no conversions, no `J...E`, no evaluating template argument expressions, etc) if either the template name is dependent, or if we're mangling an `<unresolved-qualifier-level>` or a `<base-unresolved-name>`. For everything else, we resolve as much as possible (as per the original suggestion, and as appears to be required by the language rules).

So:

The
```
template<typename T> inline typename enable_if<and_<T, T, T>::value>::type f(T) {}
```
example is mangled without a `J...E`, because `and_<T, T, T>::value` is mangled as an `<unresolved-name>` and therefore we use the source syntax for the template arguments.

In the `g1`/`g2`/`g3` example, the callee is mangled as the (unresolved) `<base-unresolved-name>` `1fIXLi0EE` in `g1`, and as the (resolved) `<expr-primary>` `_Z1fILj0EEi1XE` in the second and third cases.

One other note: in this approach, we mangle the resolved template arguments rather than the original source syntax when mangling an `<unresolved-type>`:
```
template<template<unsigned> class A> void f(typename A<0>::B) {}
template<template<unsigned> class A> void g(decltype(A<0>::C)) {}

template<unsigned> struct X { struct B {}; int C; };
void h() {
    f<X>({});
    g<X>({});
}
```
Both `f` and `g` mangle the argument as `Lj0E`, not as `Li0E` (this matches ICC, GCC, and Clang) -- in the `f` case because it's a type and therefore not an `<unresolved-name>`, and in the `g` case because it's the `<unresolved-type>` of an `<unresolved-name>`. This tweak doesn't seem necessary to follow the standard rules (`decltype(A<0u>::C)` is not equivalent to `decltype(A<0>::C)`) but is consistent with major implementations at least.
